File: ui/goth/primitives/hover_card.go
package primitives

import "github.com/a-h/templ"

// HoverCardProps configures the Hover Card container (P42, family F4): a
// supplementary, non-critical floating panel opened by hovering or focusing its
// trigger. No-JS baseline: the content is always in the DOM and revealed by a pure
// CSS hover/focus-within rule, so it works without JavaScript. The gothHoverCard
// controller enhances that baseline (it sets data-enhanced on the root, so the CSS
// follows the controller-owned data-state): hover opens after an intent delay with
// a close grace period that bridges the gap to the (possibly interactive) panel,
// focus opens immediately, and Escape, blur outside the card, or an outside press
// (touch-safe) dismisses it. It never traps focus and never moves focus into the
// panel — the card is supplementary, not a dialog. data-slot hooks: hover-card,
// trigger, hovercard-content.
type HoverCardProps struct{ Base }

// HoverCardTriggerProps configures the HoverCardTrigger. With a zero URL it
// renders a <button>; with a URL set it renders an anchor (hover cards commonly
// wrap a profile/resource link), so "links remain links".
type HoverCardTriggerProps struct {
	Base
	// URL, when set, renders the trigger as an anchor to that destination.
	URL URL
}

// HoverCardContentProps configures the HoverCardContent floating panel.
type HoverCardContentProps struct{ Base }

func hoverCardClass(p HoverCardProps) string { return classNames("goth-hover-card", p.Class) }

func hoverCardAttrs(p HoverCardProps) templ.Attributes {
	return ownedAttrs(p.Base, templ.Attributes{
		"data-slot":         "hover-card",
		"data-state":        "closed",
		"x-data":            "gothHoverCard",
		"x-on:pointerenter": "enter($event)",
		"x-on:pointerleave": "leave($event)",
		"x-on:focusin":      "focusShow($event)",
		"x-on:focusout":     "focusHide($event)",
	})
}

func hoverCardTriggerButtonAttrs(p HoverCardTriggerProps) templ.Attributes {
	return ownedAttrs(p.Base, templ.Attributes{
		"data-slot": "trigger",
		"type":      "button",
	})
}

func hoverCardTriggerLinkAttrs(p HoverCardTriggerProps) templ.Attributes {
	return ownedAttrs(p.Base, templ.Attributes{
		"data-slot": "trigger",
	})
}

func hoverCardContentAttrs(p HoverCardContentProps) templ.Attributes {
	return ownedAttrs(p.Base, templ.Attributes{
		"data-slot": "hovercard-content",
	})
}

File: ui/goth/assets/src/js/controllers/hover-card.js
// gothHoverCard controller (GOTH-4.3; added to the frozen §8 set by the recorded
// owner reopen — see gate-b-review.md 2026-07-17 addendum).
//
// Enhances a Hover Card: a supplementary, non-critical floating panel opened by
// hovering or focusing its trigger. Its no-JS baseline is the same pure CSS
// hover/focus-within reveal as Tooltip (content always in the DOM). This
// controller composes the frozen GOTH-4.1 mechanics — anchored placement
// (anchor.js) + nested-aware Escape/outside dismissal (dismiss.js/
// overlay-stack.js) — plus the shared hover-intent timing (hover-intent.js). The
// close delay bridges the gap between trigger and panel so a pointer can travel
// into the (possibly interactive) content without dismissing it. Focus opens
// immediately; Escape, blur outside the card, or an outside press (touch-safe on
// pointer devices with no hover) dismisses it. It does NOT trap focus and does not
// move focus into the panel — the card is supplementary, not a dialog. It marks
// the root data-enhanced so the CSS follows the controller-owned data-state.
// Dispatches goth:open / goth:close.

import { createAnchor } from "../mechanics/anchor.js";
import { createDismisser } from "../mechanics/dismiss.js";
import { createHoverIntent } from "../mechanics/hover-intent.js";

export default function gothHoverCard() {
  return {
    open: false,
    _root: null,
    _trigger: null,
    _content: null,
    _anchor: null,
    _dismisser: null,
    _intent: null,
    _hovering: false,
    _focused: false,

    init() {
      this._root = this.$el;
      this._trigger = this._root.querySelector('[data-slot="trigger"]');
      this._content = this._root.querySelector('[data-slot="hovercard-content"]');
      this._intent = createHoverIntent({ openDelay: 240, closeDelay: 240 });
      this._root.setAttribute("data-enhanced", "true");
    },

    enter() {
      this._hovering = true;
      this._intent.open(() => this.show());
    },

    leave() {
      this._hovering = false;
      this._scheduleClose();
    },

    focusShow() {
      this._focused = true;
      this._intent.now(() => this.show());
    },

    focusHide(event) {
      if (event && this._root.contains(event.relatedTarget)) return;
      this._focused = false;
      this._scheduleClose();
    },

    _scheduleClose() {
      if (this._hovering || this._focused) return;
      this._intent.close(() => this.hide());
    },

    show() {
      if (this.open) return;
      this.open = true;
      this._root.setAttribute("data-state", "open");
      if (this._trigger && this._content) {
        this._anchor = createAnchor(this._trigger, this._content, {
          side: "bottom",
          align: "center",
        });
      }
      this._dismisser = createDismisser(this._root, () => {
        this._hovering = false;
        this._focused = false;
        this.hide();
      });
      this.$nextTick(() => {
        if (this._anchor) this._anchor.activate();
        this._dismisser.activate();
        this.$dispatch("goth:open");
      });
    },

    hide() {
      if (!this.open) return;
      this.open = false;
      this._intent.cancel();
      this._root.setAttribute("data-state", "closed");
      if (this._dismisser) {
        this._dismisser.deactivate();
        this._dismisser = null;
      }
      if (this._anchor) {
        this._anchor.deactivate();
        this._anchor = null;
      }
      this.$dispatch("goth:close");
    },
  };
}
